### DISCUSSION TITLE
Should we encourage returning null when `id` passed to `xxxById` was not found?

### DISCUSSION THREAD

**Discussion by gojanpaolo** [ChilliCream/hotchocolate#3123] [Guidelines]
In [graphql-workshop](https://github.com/ChilliCream/graphql-workshop), the `xxxById` functions (e.g. [`SpeakerById`](https://github.com/ChilliCream/graphql-workshop/blob/de7d2c9c812429b877fdeaea82cd4fda4980fc01/code/complete/GraphQL/Speakers/SpeakerQueries.cs#L22)) defines `non-nullable` return types. This results to an error when the entity with the given `id` argument was not found.

The Relay's spec says that we should return `null` instead: https://relay.dev/graphql/objectidentification.htm#sec-Node-root-field

> The server must make a best effort to fetch this data, but it may not always be possible; for example, the server may return a `User` with a valid `id`, but when the request is made to refetch that user with the `node` root field, the user’s database may be unavailable, or the user may have deleted their account. In this case, the result of querying this field should be `null`.

Should we encourage implementation that returns `null` for such cases?

**Comment by damikun** [comment]
> In [graphql-workshop](https://github.com/ChilliCream/graphql-workshop), the `xxxById` functions (e.g. [`SpeakerById`](https://github.com/ChilliCream/graphql-workshop/blob/de7d2c9c812429b877fdeaea82cd4fda4980fc01/code/complete/GraphQL/Speakers/SpeakerQueries.cs#L22)) defines `non-nullable` return types. This results to an error when the entity with the given `id` argument was not found.
> 
> The Relay's spec says that we should return `null` instead: https://relay.dev/graphql/objectidentification.htm#sec-Node-root-field
> 
> > The server must make a best effort to fetch this data, but it may not always be possible; for example, the server may return a `User` with a valid `id`, but when the request is made to refetch that user with the `node` root field, the user’s database may be unavailable, or the user may have deleted their account. In this case, the result of querying this field should be `null`.
> 
> Should we encourage implementation that returns `null` for such cases?

All is about how you define query response... It is up on you...

The **relay** spec says if there is error return `null` and write this error to `error[]` of Graphql response.. That is basic behaviour... All is how you define types.. You can return `UserNotFound` object.. it is better representation like null in this case.. (working with response unions) Twitter is building something similar.. This pattern is common for mutation payload to return union but also useful for query response.. It needs more code and know something about unions and interfaces...

Don't take workshop as way how the program your app. This workshop is introduction to HC server and give you easy overview..  Some stuff are just dummy and simplified for people to understand quick and should not be used in that way if you wanna have robust GQL server..  So take principles from workshop...
